### ISSUE TITLE
Arrow function strict generic check issue

### ISSUE THREAD

**Issue by ibratoev** [Microsoft/TypeScript#18766]
**TypeScript Version:**  2.5.2

**Code**

```ts
interface Test {
  a: <T>() => T;
}

const t: Test = {
  a: () => 5,
};
```

**Expected behavior:**
No error
**Actual behavior:**
Error: Type 'number' is not assignable to type 'T'.

**Additional info:**
I tested with `noStrictGenericChecks` and the issue is not reproducible there.
I understand that the `T` type parameter is not really needed in this scenario. In my real scenario it is [defined like that in DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/fc018b7cbc1d7bfc6a60a4b5603f525afb50a6e4/types/react-native/index.d.ts#L5325).

**Comment by andy-ms** [comment]
You've defined an impossible function here: for *any* `T` it must return a `T`.
If you meant that for *some* `T` you must return a `T`, you can just declare the return type as `{} | null | undefined` -- since that places basically no restriction on the function at all.

**Comment by ibratoev** [comment]
@andy-ms, I understand that the definition places no restriction at all. I have linked the real case from DefinetlyTyped in the description and I am going to send a PR to fix it.

I posted the issue here because in my opinion it shouldn't cause an error. As you said, the code as it is written places no restriction but it does cause an error.

**Comment by RyanCavanaugh** [comment]
This code clearly has a type error. `t.a<string>()` is a legal invocation (per the definition of `Test`) that would observe an incorrectly-typed value. Not sure what argument there is otherwise.
